r/R/dplyr-distinct.R

```diff
@@ -0,0 +1,63 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# The following S3 methods are registered on load if dplyr is present
+
+distinct.arrow_dplyr_query <- function(.data, ..., .keep_all = FALSE) {
+  if (.keep_all == TRUE) {
+    # After ARROW-13993 is merged, we can implement this
+    arrow_not_supported("`distinct()` with `.keep_all = TRUE`")
+  }
+
+  distinct_groups <- ensure_named_exprs(quos(...))
```

Comment:
I'd expect a function body something like this:

```
.data <- dplyr::group_by(.data, ..., .add = TRUE)
gv <- dplyr::group_vars(.data) # this is `names(.data$group_by_vars)`
if (length(gv)) {
  # `data %>% group_by() %>% summarise()` returns cols in order supplied
  # but distinct() returns cols in dataset order, so sort group vars
  vars_in_data <- names(.data)[names(.data) %in% gv]
  .data$group_by_vars <- c(
    .data$group_by_vars[vars_in_data], 
    .data$group_by_vars[!(gv %in% vars_in_data)]
  )
} else {
  # distinct() with no vars specified means distinct across all cols
  .data <- dplyr::group_by(.data, !!!syms(names(.data)))
}
# TODO: use .groups = "keep" (ARROW-13550)
dplyr::summarize(.data)
```

IIUC the only specialness about distinct() over group_by() %>% summarize() is the order of columns and the meaning of 0 groups. If that's true then we should delegate the rest of the work to group_by and summarize.

Reply:
```suggestion
  .data <- dplyr::group_by(.data, ..., .add = TRUE)
  gv <- dplyr::group_vars(.data) # this is `names(.data$group_by_vars)`
  if (length(gv)) {
    # `data %>% group_by() %>% summarise()` returns cols in order supplied
    # but distinct() returns cols in dataset order, so sort group vars
    vars_in_data <- names(.data)[names(.data) %in% gv]
    .data$group_by_vars <- c(
      .data$group_by_vars[vars_in_data], 
      .data$group_by_vars[!(gv %in% vars_in_data)]
    )
  } else {
    # distinct() with no vars specified means distinct across all cols
    .data <- dplyr::group_by(.data, !!!syms(names(.data)))
  }
  # TODO: use .groups = "keep" (ARROW-13550)
  dplyr::summarize(.data)
```

Reply:
This code block is a bit hard to decipher in terms of what has or hasn't changed, and why.  I think it's complicated further by some things not quite working when the things passed into `distinct()` can be expressions, but I could be wrong, will test further.

Reply:
If it's more complicated than that, that's fine. But group_by handles expressions too, so I would expect that it would do the right thing.

Reply:
I think the nuance is in the difference between calling `group_by()` which handles expressions versus adding things to `.data$group_by_vars`, which only works (AFAIK) when it's just the names of the columns which are not expressions.  I think I could do a better job of refactoring some of this to make it a bit tidier still though - will update shortly.

Reply:
My suggestion doesn't add things to group_by_vars directly, it only reorders them to make sure they follow dataset order.

Reply:
Tbh I'm struggling to follow this a bit. I'm not sure what this bit is doing:
```
  vars_in_data <- names(.data)[names(.data) %in% gv]
  .data$group_by_vars <- c(
    .data$group_by_vars[vars_in_data], 
    .data$group_by_vars[!(gv %in% vars_in_data)]
  )
```

Reply:
Please can you add comments with what each line is doing?

Reply:
Oh, never mind, I see now. You subset to get the same thing but reordered.

Reply:
I can't get it to work that way at all, sorry. I think this is ready for re-review now @nealrichardson as have made some other changes in light of new tests which I think mean that this way no longer makes sense.

Reply:
I tried again, but I end up trying to do horrible things with lists of quosures to rearrange them and then ensure they are named, which I can't get working. Given it works the way is is written now, I'm giving up on this, unless anyone has any very specific code suggestions.

Reply:
This was annoying me, so I tried it again today.  I got 90% of the way there, but I think ultimately, it requires us to effectively undo and redo any `group_by()` which happens before the call to `distinct()` 

I ended up with this monstrosity at the start of the code:

```
gv <- dplyr::group_vars(.data)
gv_as_quo <- quos(!!!map(.data$selected_columns[gv], ~sym(.x$ToString())))
```
This isn't correct though, as it's pulling out the Expression and not the R expression used to create it, which I guess we can't get back at this point?